### PR TITLE
Add upload_to_clouds function

### PR DESCRIPTION
**Pull request by dustymabe**
This picks up #672 from @ravanelli and adds in fixups/restructure. Tested in the staging pipeline.

### CODE REVIEW

**Review by jlebon — APPROVED** [review]
Some minor comments but LGTM overall!

File: uploads.groovy [hunk 1]
```diff
@@ -7,147 +7,164 @@ def upload_to_clouds(pipecfg, basearch, buildID, stream) {
...
-                        --build=${buildID} \
-                        --region=${c.primary_region} \
-                        --bucket=s3://${c.bucket} \
-                        --credentials-file=\${ALIYUN_IMAGE_UPLOAD_CONFIG}
-                    """)
-                }
+    credentials = [file(variable: "ALIYUN_IMAGE_UPLOAD_CONFIG",
+                   credentialsId: "aliyun-image-upload-config")]
+    if (artifacts.contains("aliyun") && utils.credentialsExist(credentials)) {
+        def creds = credentials
```
[the remark this answers is not in the document]
Comment (dustymabe) [reply uploads.groovy:21]: yeah. Similar to 358d197

File: docs/config.yaml [hunk 2]
```diff
@@ -97,6 +97,19 @@ clouds:
     # REQUIRED (if GCP image upload credentials provided): the bucket+prefix
     # where image files are uploaded to then initiate an image import.
     bucket: fedora-coreos-cloud-image-uploads/image-import
+    # OPTIONAL: A description to add to the image. The given string can have ${STREAM},
+    # ${BUILDID}, ${BASEARCH}, and ${DATE} substituted for values at runtime.
+    description: "Fedora, Fedora CoreOS ${STREAM}, ${BUILDID}, ${BASEARCH} published on ${DATE}"
```
[the remark this answers is not in the document]
Comment (dustymabe) [reply docs/config.yaml:102]: I can't rememeber the exact details but this format was asked for by the GCP team when we started creating public images. I think the information is sucked into some other system.

File: uploads.groovy [hunk 3]
```diff
@@ -0,0 +1,170 @@
+
+// Upload artifacts to clouds
+def upload_to_clouds(pipecfg, basearch, buildID, stream) {
+
+    // Get artifacts to upload, if we have an artifact built
+    // try to uplaoad it
+    def meta_json = "builds/${buildID}/${basearch}/meta.json"
+    def meta = readJSON file: meta_json
```
[the remark this answers is not in the document]
Comment (dustymabe) [reply uploads.groovy:8]: It gets altered in a later commit anyway.

File: uploads.groovy [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    def artifacts = meta['images'].keySet() as List
+
+    // Define an uploader closure for each artifact/cloud that we
+    // support uploading to. Only add the closure to the map if we
+    // have credentials in our Jenkins env to support uploading to
+    // that cloud.
```
[the remark this answers is not in the document]
Comment (dustymabe) [reply uploads.groovy:14]: added some more wording

File: uploads.groovy [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    def uploaders = [:]
+    def credentials
+
+    credentials = [file(variable: "ALIYUN_IMAGE_UPLOAD_CONFIG",
+                   credentialsId: "aliyun-image-upload-config")]
```
[the remark this answers is not in the document]
Comment (dustymabe) [reply uploads.groovy]: fixed

File: uploads.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (dustymabe) [reply uploads.groovy:8]: actually I just realized there were a few commits i thought were in this batch that weren't. Just pushed two new commits.

**Comment by dustymabe** [comment]
Added two commits here that I thought were included in this PR. @jlebon - do you mind taking at look at those two new commits?

File: uploads.groovy [hunk 6]
```diff
@@ -2,11 +2,9 @@
 // Upload artifacts to clouds
 def upload_to_clouds(pipecfg, basearch, buildID, stream) {
 
-    // Get artifacts to upload, if we have an artifact built
-    // try to uplaoad it
-    def meta_json = "builds/${buildID}/${basearch}/meta.json"
-    def meta = readJSON file: meta_json
-    def artifacts = meta['images'].keySet() as List
+    // Get a lits of the artifacts that are currently built.
```
Comment (jlebon) [inline uploads.groovy:5]: Minor: list

**Review by jlebon — APPROVED** [review]

File: uploads.groovy (hunk 6 above)
Comment (dustymabe) [reply uploads.groovy]: fixed!

**Pull request merged**